### ISSUE TITLE
Need mp2/mp3 creation

### ISSUE THREAD

**Issue by brettviren** [WireCell/spng#9]
With #6 done, we next need to use ray grid tiling to provide "mp2/mp3" images to dnnroi.

There are a few ways this can be structured at the DFP graph level.  My initial thoughts are shown in the following graph.  These are just meant to provide some initial guidance and not necessarily the best way to do things in the end.  Some description and comments follow the graph.

<img width="429" height="1211" alt="Image" src="https://github.com/user-attachments/assets/e1d34f80-70d4-4db1-b684-000f666ecd1f" />

- Boxes are DFP graph nodes
- Bubbles are `ITorchTensorSet`'s
- The `rebin` node should perform a time-downsampling consistent with the signal bandwidth.  This is typically rebin-4.
- The `InitialROIFinder` is likely the already ported initial ROI finding and minus the "roi refinement".  It must be good enough to handle the amplified noise.  
- We may need to bifurcate the start of this graph to include "wiener+looseLF" and "wiener+tightLF" as in OSP.
- The `mpx_*` are two instances of one component that makes either mp2 or mp3 type images.  I split them assuming it is easiest to provide a single implementation with configuration-driven variant behavior. 
- This graph is the union of two separate graphs with different outputs: one for producing training data (needs unpictured graph: depo->splat->threshold to make roi truth) and one for feeding pre-trained inference.

The `tiling` node:
- Runs ray grid tiling and outputs a "blobs tensor" for each time slice and each APA face
  - Shape `(nblobs, nviews, 2)` giving half-open bounds in each tomographic view in terms of ray (wire) indices.
- Tiling is a per-FACE and per-SLICE (rebin-4 "tick") operation.
  - Maybe want to break the tiling "node" into a subgraph of parts, 
    - initial node to split by FACE
    - splitting by SLICE is probably too insane (1000s of nodes needed)
  - In any case, the output "blobs" `ITorchTensorSet` will have one `ITorchTensor` holding all blobs for one time slice and one face.

**Comment by calcuttj** [comment]
I think we're mostly there. I've added a step to take us back from MP2/3 per each wire segment collapsed onto channel. 

Similar to the 1:1 Wire segment to channel map, I added a 1:many channel to wire segment map 
https://github.com/WireCell/spng/blob/9eac17dd9dddde9cea5af93258bf3e496db8df7b/src/FindMPCoincidence.cxx#L144-L161

I give a dummy channel number that is later always set to 0/low

Then we use torch::gather to use the channel->Wire mapping as indices for the wire segment MP2/3. This is repeated for every time sample.
https://github.com/WireCell/spng/blob/9eac17dd9dddde9cea5af93258bf3e496db8df7b/src/FindMPCoincidence.cxx#L180-L192

Here are the outputs

MP3:
<img width="1629" height="864" alt="Image" src="https://github.com/user-attachments/assets/52908abb-a48a-4010-baca-8f3595ef0b3a" />

MP2:
<img width="1629" height="864" alt="Image" src="https://github.com/user-attachments/assets/45dd0bfd-fe0a-47fe-a646-594467533ffe" />

Axes are Electronics Channel vs Ticks Rebinned by 4

For MP3, we see clearly one track (I believe the channels at the bottom are from wrapping),  this requires coincidence activity for all 3 planes, so except for some noise, we see the track which kind of by construction we know should be the only MP3.

For MP2, the interpretation is a bit trickier. Basically it shows any channel for which any of its wire segments were low AND over which active wires from the other 2 planes cross. We can clearly see where MP3 was in the low region between two high regions. I think this is from crossings of other-plane active wires on nearby wires in the target plane.

**Comment by calcuttj** [comment]
Yes there is noise

**Comment by brettviren** [comment]
Is this pattern also in MP3 but just not visible due to a color scale difference?  If only in mp2, what would be the cause?

**Comment by calcuttj** [comment]
What's drawn is binary (0/1) but matplotlib renders it a bit weird. So here they are in gray scale

<img width="1629" height="864" alt="Image" src="https://github.com/user-attachments/assets/73cbe878-70c8-4681-bcd3-ad5dcc7634af" />

<img width="1629" height="864" alt="Image" src="https://github.com/user-attachments/assets/17fb4adf-641a-428b-9f5d-5739a49fb5f3" />

So the pattern isn't there in MP3. I'll dig into/think about the pattern in MP2

**Comment by calcuttj** [comment]
Yeah the current ROI implementation is quite brute force and zeros out anything less than some factor times the RMS per wire. Raising it from 3 to 4.5 gives this 
MP3:
<img width="1629" height="864" alt="Image" src="https://github.com/user-attachments/assets/419e818a-8e6e-4c37-ad05-17663c57017e" />

MP2:
<img width="1629" height="864" alt="Image" src="https://github.com/user-attachments/assets/efd38fd5-bc6c-43f3-b902-3b96085724af" />

**Comment by calcuttj** [comment]
I could run the script to make the efficiency & purity vs angle plots to see this. 

I agree on the full chain, though I might actually try doing the simpler MP2/MP3 creation that doesn't use tiling.

**Comment by brettviren** [comment]
Yes, at some point we will need to run the metric-vs-angle test.  But a reasonable first goal is full chain that outputs something that is somewhat okay "by eye" on a single event holding a single ideal track.  We can then refine the pipeline.

In addition to pursuing non-tiling mp2/mp3, the other option is a better trained, 3-plane DNNROI that inputs only Gauss filtered decon images.  But, again, I suggest we complete the initial imperfect full chain first.

**Comment by calcuttj** [comment]
> > I think we can keep the batching of the MP2/MP3 data coming out of this node when we send through DNNROI
> 
> Do you mean the flattened `NB*NS` here? If so, I suggest we always restore the full `(NB,NS,NC)` shape on output. This kind of reshaping is fast and by doing it, it allows for more consistent assumptions through the graph.
> 

I mean the NB, NS, NC shape

**Comment by brettviren** [comment]
A `CrossViews` node is added that computes "MP0" and "MP1" in addition to "MP2" and "MP3" and encodes these per-wire values into per-channel packed integers.  This node is essentially an extended version of Jake's NoTileMPCoincidence node.  Between the two of them SPNG has this issue closed.
